File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {

        int prefixSum = 0 ;
        unordered_map<int , int> mpp;

        int count = 0 ;
        mpp[0] = 1;

        for(int i = 0 ; i < nums.size() ;i++){

            prefixSum += nums[i] ;

            auto it = prefixSum - goal ;

            if(mpp.find(it) != mpp.end()){
                count += mpp[it];
            }

            mpp[prefixSum]++;

        }

        return count ;

    }
};

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {

    public int helper(int[] nums, int goal) {

        if(goal < 0)return 0 ; 

        int total = 0 ;
        int start = 0 ;
        int sum = 0 ;

        for(int i = 0 ; i < nums.length ; i++){

            sum += nums[i];
            
            while(start <= i && sum > goal){
                sum -= nums[start++];
            }

            total += (i - start + 1);
            
        }
        
        
        return total;
        
    }


    public int numSubarraysWithSum(int[] nums, int goal) {

        return helper(nums , goal) - helper(nums , goal - 1) ; 



    }
}